File: php/05.php
<?php

$inputFile = '../inputs/input05.txt';
$data = file($inputFile);

$data = array_map(function($line) {
    preg_match("/(?<x1>\d+),(?<y1>\d+) -> (?<x2>\d+),(?<y2>\d+)/", $line, $m);
    $matches = [$m['x1'], $m['y1'], $m['x2'], $m['y2']];
    return array_map(
        function ($x) { return intval($x); },
        $matches
    );
}, $data);

function expand_line($coords, $opts) {
    [$x1, $y1, $x2, $y2] = $coords;

    $xrange = range($x1, $x2);
    $yrange = range($y1, $y2);

    // horizontal
    if ($x1 == $x2) {
        return array_map(
            function ($y) use ($x1) { return [$x1, $y];},
            $yrange
        );
    }
    // vertical
    else if ($y1 === $y2) {
        return array_map(
            function ($x) use ($y1) { return [$x, $y1];},
            $xrange
        );
    }
    // diagonals allowed
    else if ($opts['allow_diagonals']) {
        $zipped = [];
        for ($i = 0; $i < count($xrange); $i++) {
            $zipped[] = [$xrange[$i], $yrange[$i]];
        }
        return $zipped;
    }
    // diagonals ignored
    else {
        return [];
    }
}

function count_hotspots($data, $opts) {
    $expanded_lines = array_map(
        function($line) use ($opts) { return expand_line($line, $opts); },
        $data
    );
    $expanded_points = array_merge(...$expanded_lines);
    $point_strings = array_map(
        function ($point) { return implode('_', $point); },
        $expanded_points
    );
    $counts = array_count_values($point_strings);
    $hotspots = array_filter(
        $counts,
        function($v) { return $v > 1; }
    );
    return count($hotspots);
}

// part 1
$n1 = count_hotspots($data, ['allow_diagonals' => false]);
print "P1: $n1\n";

// part 2
$n2 = count_hotspots($data, ['allow_diagonals' => true]);
print "P2: $n2\n";

File: elixir/05.exs
defmodule Day05 do
  defp read_input do
    File.read!(Path.expand("../inputs/input05.txt"))
    |> String.split("\n")
    |> Enum.map(&regex_parse_line/1)
  end

  defp regex_parse_line(line) do
    captures =
      Regex.named_captures(
        ~r/(?<x1>\d+),(?<y1>\d+) -> (?<x2>\d+),(?<y2>\d+)/,
        String.trim(line)
      )

    %{
      :x1 => captures["x1"] |> String.to_integer(),
      :y1 => captures["y1"] |> String.to_integer(),
      :x2 => captures["x2"] |> String.to_integer(),
      :y2 => captures["y2"] |> String.to_integer()
    }
  end

  defp expand_line(%{x1: x1, y1: y1, x2: x2, y2: y2} = _coords, opts) do
    xrange = x1..x2
    yrange = y1..y2

    cond do
      # horizontal
      x1 == x2 ->
        for y <- yrange, do: {x1, y}

      # vertical
      y1 == y2 ->
        for x <- xrange, do: {x, y1}

      # diagonals allowed
      Keyword.get(opts, :allow_diagonals) ->
        Enum.zip(xrange, yrange)

      # diagonals ignored
      true ->
        []
    end
  end

  defp count_hotspots(opts) do
    # use {x,y} tuples as keys in a counter Map
    read_input()
    |> Enum.map(fn line -> expand_line(line, opts) end)
    |> Enum.flat_map(fn x -> x end)
    |> Enum.reduce(
      %{},
      fn point, acc ->
        Map.update(acc, point, 1, &(&1 + 1))
      end
    )
    |> Map.values()
    |> Enum.filter(fn x -> x > 1 end)
    |> length
  end

  @doc """
  Count the number of grid cells containing 2 or more line parts - diagonals excluded
  """
  def part1 do
    count_hotspots(allow_diagonals: false)
    |> IO.inspect(label: "P1")
  end

  @doc """
  Count the number of grid cells containing 2 or more line parts - diagonals included
  """
  def part2 do
    count_hotspots(allow_diagonals: true)
    |> IO.inspect(label: "P2")
  end
end

# P1: 4826
# P2: 16793
